src/main/scala/chisel3/util/Queue.scala

```diff
@@ -185,6 +188,87 @@ object Queue {
     }
   }
 
+  /** A factory for creating shadow queues.  This is created using the
+    * `withShadowQueue` method.
+    */
+  class ShadowFactory[T <: Data] private[Queue] (
+    enq:            ReadyValidIO[T],
+    deq:            ReadyValidIO[T],
+    entries:        Int,
+    pipe:           Boolean,
+    flow:           Boolean,
+    useSyncReadMem: Boolean,
+    flush:          Option[Bool]) {
+
+    /** Create a "shadow" `Queue` in a specific layer that will be queued and
+      * dequeued in lockstep with an original `Queue`.  Connections are made
+      * using `BoringUtils.tapAndRead` which allows this method to be called
+      * anywhere in the hierarchy.
+      *
+      * An intended use case of this is as a building block of a "shadow" design
+      * verification datapath which augments an existing design datapath with
+      * additional information.  E.g., a shadow datapath that tracks transations
+      * in an interconnect.
+      *
+      * @param data a hardware data that should be enqueued together with the
+      * original `Queue`'s data
+      * @param layer the `Layer` in which this queue should be created
+      * @return a layer-colored `Valid` interface of probe type
+      */
+    def apply[A <: Data](data: A, layer: Layer): Valid[A] = {
+      val shadowDeq = Wire(Probe(Valid(chiselTypeOf(data)), layer))
+
+      block(layer) {
+        val shadowEnq = Wire(Decoupled(chiselTypeOf(data)))
+        shadowEnq.ready :<= true.B
+        shadowEnq.valid :<= BoringUtils.tapAndRead(enq).fire
+        shadowEnq.bits :<= data
+
+        val shadowQueue = Queue(shadowEnq, entries, pipe, flow, useSyncReadMem, flush.map(BoringUtils.tapAndRead))
+
+        val _shadowDeq = Wire(Valid(chiselTypeOf(data)))
+        _shadowDeq.valid :<= shadowQueue.valid
```

Comment:
Should we assert that this valid always equals the original valid...? Is there a reason they should not?

Reply:
These should always be the same. If they're not then it's a bug in this generator.

Maybe we want something that is a layer for the purposes of testing the generator itself. 🧐 I'm going to take no action on this for now.